test/xbgpu/test_bsend.py

```diff
@@ -0,0 +1,190 @@
+
+    @staticmethod
+    async def _recv_data(
+        recv_stream: spead2.recv.asyncio.Stream,
+        channel_offset: int,
+        n_channels_per_substream: int,
+        n_spectra_per_heap: int,
+        heap_timestamp_step: int,
+    ) -> None:
+        """Receive data transmitted from :func:`_send_data`.
+
+        Error-check data here as well.
+        """
+        ig = spead2.ItemGroup()
+
+        # Wait for the first packet to arrive - it is expected to be the
+        # SPEAD descriptor.
+        heap = await recv_stream.get()
+        # TODO: No checks on the count sequence yet
+        items = ig.update(heap)
+        assert items == {}, "This heap contains item values not just the expected descriptors."
+
+        # Check the data heaps
+        zero_data = np.zeros(shape=(n_channels_per_substream, n_spectra_per_heap, COMPLEX), dtype=np.int8)
+        for i in range(TOTAL_HEAPS):
+            heap = await recv_stream.get()
+            items = ig.update(heap)
+            assert set(items.keys()) == {"timestamp", "frequency", "bf_raw"}
+            assert items["timestamp"].id == TIMESTAMP_ID
+            assert items["timestamp"].value == i * heap_timestamp_step
+            assert items["frequency"].id == FREQUENCY_ID
+            assert items["frequency"].value == channel_offset
+            assert items["bf_raw"].id == BF_RAW_ID
+            assert items["bf_raw"].value.shape == (n_channels_per_substream, n_spectra_per_heap, COMPLEX)
+            assert items["bf_raw"].value.dtype == np.int8
+            np.testing.assert_equal(items["bf_raw"].value, zero_data)
+
+    @pytest.mark.parametrize("num_channels", test_parameters.num_channels)
+    @pytest.mark.parametrize("num_spectra_per_heap", test_parameters.num_spectra_per_heap)
+    async def test_send_simple(
+        self,
+        context: AbstractContext,
+        num_channels: int,
+        num_spectra_per_heap: int,
+        output: BOutput,
+        time_converter: TimeConverter,
+        sensors: SensorSet,
+    ) -> None:
+        """
+        Test :class:`katgpucbf.xbgpu.bsend.BSend`.
+
+        This test transmits a number of heaps from a BSend object over a spead2
+        in-process transport. The received heaps are then checked.
+
+        This test does not generate random data as it will take much more compute
+        to check that the random data is received correctly.
+
+        Parameters
+        ----------
+        context
+            Device context for allocating buffers.
+        num_channels
+            Total number of channels processed by a (theoretical) F-engine.
+        num_spectra_per_heap
+            Total number of packed spectra in every recevied channel.
+        """
+        # TODO: We don't do channels * 2 anymore, but n-samples-between-spectra
+        heap_timestamp_step = num_channels * 2 * num_spectra_per_heap
+        # Arbitrarily chosen, channels-per-substream is dictated by the
+        # F-engine anyway.
```

Comment:
Well, it's dictated by the number of XB-engines that we need for a given n_antennas.
It's passed as a parameter to the F-engine.

Reply:
I'd still suggest rewording this comment to something like
```suggestion
        # Arbitrarily chosen, channels-per-substream is dictated by the
        # top-level correlator configuration anyway.
```